### PR TITLE
ZCS-2389:Modification of AuthRequest to support JWT

### PR DESCRIPTION
**Pull request by shrikant-prasad**
Added optional attribute "tokenType" in AuthRequest.
If tokenType is 'JWT', jwt will be returned in response, otherwise auth token will be returned.
Added ldap attributes to store jwt key and jwt lifetime.
Added junit test.

### CODE REVIEW

File: store/src/java/com/zimbra/cs/service/account/Auth.java [hunk 1]
```diff
@@ -118,6 +120,12 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
         }
 
         String password = request.getAttribute(AccountConstants.E_PASSWORD, null);
+        String tokenType = request.getAttribute(AccountConstants.A_TOKEN_TYPE, null);
+        if (Constants.JWT_TOKEN_TYPE.equalsIgnoreCase(tokenType)) {
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/account/Auth.java]: Why not just `isJWT = Constants.JWT_TOKEN_TYPE.equalsIgnoreCase(tokenType)`

File: store/src/java/com/zimbra/cs/service/account/Auth.java [hunk 2]
```diff
@@ -446,24 +485,9 @@ private Element doResponse(Element request, AuthToken at, ZimbraSoapContext zsc,
         String skin = SkinUtil.chooseSkin(acct, requestedSkin);
         ZimbraLog.webclient.debug("chooseSkin() returned "+skin );
         if (skin != null) {
-            response.addElement(AccountConstants.E_SKIN).setText(skin);
+            response.addUniqueElement(AccountConstants.E_SKIN).setText(skin);
```
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/account/Auth.java]: This looks like a change in the API and may break things.

File: store/src/java/com/zimbra/cs/util/AccountUtil.java [hunk 3]
```diff
@@ -636,4 +652,50 @@ public static int getRootFolderIdForItem(MailItem item, Mailbox mbox,
...
+        long lifeTime = acct.getTimeInterval(Provisioning.A_zimbraJWTLifetime, DEFAULT_JWT_LIFETIME * 1000);
+        long now = System.currentTimeMillis();
+        long expires = now + lifeTime;
+        builder.setIssuedAt(new Date(now));
+        builder.setExpiration(new Date(expires));
+        String compactJws = builder.signWith(signatureAlgorithm, key).compact();
+        return compactJws;
+    }
+
+    private static synchronized String genearteJWTkey(String tokenKey, Account acct) throws ServiceException {
```
Comment (tmclane) [inline store/src/java/com/zimbra/cs/util/AccountUtil.java]: Please fix this typo.
genearteJWTKey => generateJWTKey

File: store/src/java/com/zimbra/cs/util/AccountUtil.java [hunk 4]
```diff
@@ -636,4 +652,50 @@ public static int getRootFolderIdForItem(MailItem item, Mailbox mbox,
         }
         return subscriptions;
     }
+
+    public static String generateJWT(Account acct, boolean isAdmin) throws ServiceException {
+        io.jsonwebtoken.SignatureAlgorithm signatureAlgorithm = io.jsonwebtoken.SignatureAlgorithm.HS512;
+        String tokenKey = acct.getJWTKey();
+        if (StringUtil.isNullOrEmpty(tokenKey)) {
+            tokenKey = genearteJWTkey(tokenKey, acct);
```
Comment (tmclane) [inline store/src/java/com/zimbra/cs/util/AccountUtil.java]: Please fix this typo.
genearteJWTKey => generateJWTKey

File: store/src/java/com/zimbra/cs/util/AccountUtil.java (hunk 3 above)
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/util/AccountUtil.java]: done.

File: store/src/java/com/zimbra/cs/util/AccountUtil.java (hunk 4 above)
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/util/AccountUtil.java]: done.

File: store/src/java/com/zimbra/cs/service/account/Auth.java (hunk 2 above)
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/account/Auth.java]: reverted to addElement, only reason for change was that addElement has been deprecated.

File: store/src/java/com/zimbra/cs/service/account/Auth.java (hunk 1 above)
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/account/Auth.java]: done.

**Comment by rohan-ambasta** [comment]
Please explore how we can extend the current AuthToken and AuthProvider for JWT

File: store/src/java/com/zimbra/cs/service/account/Auth.java (hunk 2 above)
Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/account/Auth.java]: Please change it to `addNonUniqueElement`.  That is the non-deprecated method call which does exaclty the same thing as `addElement` does - as mentioned in the JavaDoc for `addElement` explaining why it was deprecated.

Comment (fatkudu) [inline store/src/java/com/zimbra/cs/service/account/Auth.java]: I've added ZCS-2498 a Jira suggesting a global search and replace to change all addElement calls to addNonUniqueElement.  If I can get away with it, will delete addElement as well.

Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/account/Auth.java]: done.

**Comment by plobbes** [comment]
I've added comments to related JIRA tickets, but wanted to be sure we cover the topic here too. Along with this implementation change/addition to ZCS, we need to document how we are protecting from CSRF abuse while introducing these changes.  These may be useful references:
- https://stackoverflow.com/questions/27067251/where-to-store-jwt-in-browser-how-to-protect-against-csrf
- https://stackoverflow.com/questions/21357182/csrf-token-necessary-when-using-stateless-sessionless-authentication

File: store/src/java/com/zimbra/cs/service/AuthProvider.java [hunk 5]
```diff
@@ -516,6 +534,32 @@ public static AuthToken getAuthToken(Account acct) throws AuthProviderException
...
 
+    public static AuthToken getAuthToken(Account acct, TokenType tokenType) throws AuthProviderException {
+        List<AuthProvider> providers = getProviders();
+        AuthProviderException authProviderExp = null;
+
+        for (AuthProvider ap : providers) {
+            try {
+                AuthToken at = ap.authToken(acct, tokenType);
+                if (at == null) {
+                    authProviderExp = AuthProviderException.FAILURE("auth provider " + ap.getName() + " returned null");
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/AuthProvider.java]: fixed.

File: store/src/java/com/zimbra/cs/service/AuthProvider.java [hunk 6]
```diff
@@ -516,6 +534,32 @@ public static AuthToken getAuthToken(Account acct) throws AuthProviderException
...
+                } else {
+                    return at;
+                }
+            } catch (AuthProviderException e) {
+                if (e.canIgnore()) {
+                    logger().debug(ap.getName() + ":" + e.getMessage());
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/AuthProvider.java]: fixed

File: store/src/java/com/zimbra/cs/service/AuthProvider.java [hunk 7]
```diff
@@ -581,6 +625,32 @@ public static AuthToken getAuthToken(Account acct, long expires) throws AuthProv
...
+            try {
+                AuthToken at = ap.authToken(acct, expires, tokenType);
+                if (at == null) {
+                    authProviderExp = AuthProviderException.FAILURE("auth provider " + ap.getName() + " returned null");
+                } else {
+                    return at;
+                }
+            } catch (AuthProviderException e) {
+                if (e.canIgnore()) {
+                    logger().debug(ap.getName() + ":" + e.getMessage());
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/AuthProvider.java]: fixed

**Comment by shrikant-prasad** [comment]
@plobbes @Prashantsurana for these issue, we have bug already open. ZCS-2504.

**Comment by plobbes** [comment]
Since we're leaving security for this to ZCS-2504, we shouldn't commit this until we address ZCS-2504. We shouldn't introduce (even "temporary") security holes in the product.

**Comment by shrikant-prasad** [comment]
These changes will be committed to feature branch as you can see from pull request.  It will be merged to develop branch (product) only when once all issues are fixed.You can't expect all changes to be done under single ticket.

**Comment by plobbes** [comment]
As long as this change doesn't make it outside the feature branch before ZCS-2504 we're fine.  That was my intent, although perhaps I needed to be more specific with my wording.  Sorry about that.

**Pull request closed without merging**